pkg/execution/execution.go

```diff
@@ -23,55 +23,116 @@ package execution
 import (
 	"context"
 	"errors"
+	"sort"
 	"time"
 
 	"github.com/dop251/goja"
 
 	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
 	"go.k6.io/k6/lib"
 )
 
-// Execution is a JS module to return information about the execution in progress.
-type Execution struct{}
+type (
+	// RootModule is the global module instance that will create module
+	// instances for each VU.
+	RootModule struct{}
 
-// New returns a pointer to a new Execution.
-func New() *Execution {
-	return &Execution{}
+	// ModuleInstance represents an instance of the execution module.
+	ModuleInstance struct {
+		modules.InstanceCore
+		proxy *goja.Proxy
+	}
+)
+
+var (
+	_ modules.IsModuleV2 = &RootModule{}
+	_ modules.Instance   = &ModuleInstance{}
+)
+
+// New returns a pointer to a new RootModule instance.
+func New() *RootModule {
+	return &RootModule{}
 }
 
-// GetVUStats returns information about the currently executing VU.
-func (e *Execution) GetVUStats(ctx context.Context) (goja.Value, error) {
-	vuState := lib.GetState(ctx)
-	if vuState == nil {
-		return nil, errors.New("getting VU information in the init context is not supported")
+// NewModuleInstance implements the modules.IsModuleV2 interface to return
+// a new instance for each VU.
+// It initializes a goja.Proxy instance, which in turn returns
+// goja.DynamicObject instances for each property (scenario, vu, test).
+func (*RootModule) NewModuleInstance(m modules.InstanceCore) modules.Instance {
+	keys := []string{"scenario", "vu", "test"}
+
+	pcfg := goja.ProxyTrapConfig{
+		OwnKeys: func(target *goja.Object) *goja.Object {
+			rt := m.GetRuntime()
+			return rt.ToValue(keys).ToObject(rt)
+		},
+		Has: func(target *goja.Object, prop string) (available bool) {
+			return sort.SearchStrings(keys, prop) != -1
+		},
+		Get: func(target *goja.Object, prop string, r goja.Value) goja.Value {
+			return dynObjValue(m.GetContext, target, prop)
+		},
+		GetOwnPropertyDescriptor: func(target *goja.Object, prop string) (desc goja.PropertyDescriptor) {
+			desc.Enumerable, desc.Configurable = goja.FLAG_TRUE, goja.FLAG_TRUE
+			desc.Value = dynObjValue(m.GetContext, target, prop)
+			return desc
+		},
```

Comment:
Any reason why you went with a Proxy instead of a simple Object + accessor functions?

Reply:
To elaborate I would've:
1. made golang type wrapping modules.InstanceCore and having simple getMethods as before
2. made a goja object and set the appropriate names as accessor that call the get functions